storagenode/vouchers/service.go

```diff
@@ -7,16 +7,165 @@ import (
 	"storj.io/storj/pkg/pb"
 	"storj.io/storj/pkg/storj"
+	"storj.io/storj/pkg/transport"
+	"storj.io/storj/storagenode/trust"
+)
+
+var (
+	// VoucherError represents errors with vouchers
+	VoucherError = errs.Class("voucher")
+
+	mon = monkit.Package()
 )
 
 // DB implements storing and retrieving vouchers
 type DB interface {
 	// Put inserts or updates a voucher from a satellite
 	Put(context.Context, *pb.Voucher) error
-	// GetExpiring retrieves all vouchers that are expired or about to expire
-	GetExpiring(context.Context, time.Duration) ([]storj.NodeID, error)
 	// GetValid returns one valid voucher from the list of approved satellites
 	GetValid(context.Context, []storj.NodeID) (*pb.Voucher, error)
+	// NeedVoucher returns true if a voucher from a particular satellite is expired, about to expire, or does not exist
+	NeedVoucher(context.Context, storj.NodeID, time.Duration) (bool, error)
+}
+
+// Config defines configuration for requesting vouchers.
+type Config struct {
+	Interval         int `help:"number of days between voucher service iterations" default:"7"`
+	ExpirationBuffer int `help:"buffer period of X days into the future. If a voucher would expire within this period, send a request for renewal" default:"7"`
+}
+
+// Service is a service for requesting vouchers
+type Service struct {
+	log *zap.Logger
+
+	kademlia  *kademlia.Kademlia
+	transport transport.Client
+
+	vdb   DB
+	trust *trust.Pool
+
+	expirationBuffer time.Duration
+
+	Loop sync2.Cycle
+}
+
+// NewService creates a new voucher service
+func NewService(log *zap.Logger, kad *kademlia.Kademlia, transport transport.Client, vdb DB, trust *trust.Pool, interval, expirationBuffer time.Duration) *Service {
+	return &Service{
+		log:              log,
+		kademlia:         kad,
+		transport:        transport,
+		vdb:              vdb,
+		trust:            trust,
+		expirationBuffer: expirationBuffer,
+		Loop:             *sync2.NewCycle(interval),
+	}
+}
+
+// Run sends requests to satellites for vouchers
+func (service *Service) Run(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	return service.Loop.Run(ctx, service.RunOnce)
+}
+
+// RunOnce runs one iteration of the voucher request service
+func (service *Service) RunOnce(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	service.log.Info("Checking vouchers")
+
+	satellites := service.trust.GetSatellites((ctx))
+
+	if len(satellites) > 0 {
+		var group errgroup.Group
+		ctx, cancel := context.WithTimeout(ctx, time.Hour)
+		defer cancel()
+		for _, sat := range satellites {
+			sat := sat
+			need, err := service.vdb.NeedVoucher(ctx, sat, service.expirationBuffer)
+			if err != nil {
+				service.log.Error("getting voucher status", zap.Error(err))
+				return nil
+			}
+			if need {
+				group.Go(func() error {
+					service.Request(ctx, sat)
+					return nil
+				})
+			}
+		}
+		_ = group.Wait() // doesn't return errors
+	} else {
+		service.log.Debug("No voucher requests to send")
+	}
```

Comment:
can we change this to check `if len(satellites) == 0; service.log.Debug("No voucher requests to send"); return nil` first, then we dont need the rest indented

Reply:
im running storj-sim right now and its not clear what this message means

```
DEBUG	vouchers	vouchers/service.go:107	No voucher requests to send
```
Maybe we could add something like, "No trusted satellites so not requesting any vouchers to approve storing data".  What do you think?